### DISCUSSION TITLE
Any solution for combining different types of workflows (fast bashing leading into detailed upscalers) into one workflow?

### DISCUSSION THREAD

**Discussion by unphased** [Comfy-Org/ComfyUI#2513] [Q&A]
Got a question for you experts. Now I've been messing around and I know my way around but I find that there is a problem with scale and I want to ask what the best solutions are today for dealing with this issue of reconciling two types of workflow because the workflows feed into each other. So far though it seems like we have to manually manage chosen images and import them across separate workflows. 

So in the course of generating images:

First, we want to use a high speed (turbo & LCM SDXL, or 1.5 with LCM Lora) workflow to generate many images at multiple  images per second. 

Then, I would like to have some sort of a picker node where I can click in the gallery of outputs any that I like, that I'll spend more compute feeding through: detailer flows, or maybe send through a non-LCM checkpoint img2img flow, then optionally detailers again, optionally upscalers, etc. 

Basically we have separate workflows going on, there's the initial composition bashing step (mode A: rapid generation) where we can generate hundreds of outputs per minute, in large batches. 

Then there is detailed and could be extremely complex flows to work off of the compositions we found that we like as starting points. These (mode B: detailed rendering) flows take much longer to run. 

The conflict is that I have not seen an effective method to switch a comfy workflow from mode A into mode B. 

- We have to toggle seed increment/randomize off to fixed in order to safely click Generate without blasting out more generations at the front of the chain and invalidating everything else. This takes about 10s as we have to hunt down that node and flip a setting there.
- We have to drag around noodles based on a plan that we come up with for how we want to enhance the image we chose. For example say we got one LCM gen we like out of this most recent batch of 16, I dunno, I assume a custom node exists where we can view all 16 and click within it to pick a specific desired image, and feed it downstream. OK, then downstream, if I want to first do img2img through a checkpoint, then feed it through detailer, then feed it through an upscale, and a detailer again, assuming those are all pre-prepared as modular node group flows, I will spend about a minute on wiring up at least 4 noodles to connect this sequence of sequential operations (that's optimistic because many of these modules will need multiple inputs but I'll also hand wave it as we can have custom pipe nodes to help us collapse their connections into single ones to help us out).
- The flow from this point on probably involves rewiring the different modules, maybe heavy fiddling with settings in detailers and upscalers, etc. It's pretty efficient since seed has been fixed so each time we Generate only what changed needs to be recomputed.

OK got a good result after screwing around, Now to return from mode B to mode A:

- Toggle seed increment back to increment or randomize in initial ksampler
- Sever the connection from gallery output to detailed rendering flows. Reconnect output to Save Images or whatever
- Now we can begin generating again

Now that I have constructed an overview of how this would work in the ideal case maybe I'm blowing the manual overhead out of proportion. But I really do feel like it's really tedious to switch back and forth. And I start to feel like the behavior of the Generate button and how it manipulates the seed needs some sort of work. Basically I think what we should have is the notion that some nodes (such as the hypothetical gallery image picker I described above) can be made to auto trigger execution down its nodes when manipulated. I am not sure but I suspect this capability is not possible because the Generate button must be clicked for any node evaluation to proceed. 

This way, we could keep the generate button dedicated for compo bashing while the magic gallery image picker is the entry point for detailer flows coming off of it. It should be a custom node that is an image picker that also has a generate button that applies to the flows downstream.

The alternative to all of this would be instead of trying to do it all in one workflow just have them as separate workflows. I just don't know how to make this efficient but the idea is to just make larger compobashing batches, choose the ones we want to enhance further using an out of band tool, and then come up with a way with script or manually load the chosen ones as inputs for a completely separate workflow. It's probably a more straightforward approach for attaining efficiency but I think it requires far more discipline than I'll have. I'm gonna be too curious about what i can achieve with some gens as soon as I see them. Since it's such a creative process, the order of operations matters too much here to dismiss out of hand. 

Another alternative is to have automations or custom nodes or something that can let us automate all those mode switching steps down into one trigger somehow. So we could have special noodles or special pairs of nodes that effectively implement "automated noodles".

**Comment by Proryanator** [comment]
I'd be interested in something like this as well, although I might explore programmatically creating something like this.

Given 2 workflows, be able to programmatically combine the two:

- add the nodes from workflow 2 into workflow 1 (accounting for/adjusting for any naming conflicts if any)
- this should be possible via reading the .json of the first workflow and looking at the "order" value to find which is the last node
- whichever node is the last one would need to support/have an output capability (maybe a dynamic swap out of a SaveImage node in favor of some other sort of passthrough node)
- would need to modify the "links" value of the last node in the first workflow to hook up to the first node of the next workflow (assuming that workflow accepts an image input and not a prompt for example)
- make any other adjustments, including potentially tacking on a SaveImage node to the end of the combined workflow

There would be a lot of assumptions but, assuming the author of the workflows (and the tool to combine them) was aware of these things, might be doable.

If I end up building something that can do this I'll share it with you! Most posts I've found online say copy-paste, it's just that if at any point you need to update/modify the workflows separately you now need to re-do copy-pasting in any combined workflows which is a pain to manage.

**Reply by unphased** [reply]
Cool, also btw I haven't been tinkering in comfy or keeping up with it for a while, but my gut tells me there are already tons of custom nodes that we could use for various types of "switching" and other clever stuff.

**Reply by Proryanator** [reply]
Yes that's what I noticed too, there's a lot of good available workflows ready to use!
